bpo-32107 - Fix test_uuid

warsaw opened the pull request:
The original check was erroneous in two ways.  First, the documentation said
that "47 bit will never be set in IEEE 802 addresses obtained from network
cards".  I think this is referring to the universally vs. locally administered
MAC addresses.  Network cards from hardware manufacturers will always be
universally administered in order to guarantee global uniqueness.

But from https://en.wikipedia.org/wiki/MAC_address it says that this flag is
"distinguished by setting the second-least-significant bit of the first
octet of the address", where a 0 means universally administered and a 1 means
it's locally administered.  This works out to the 42nd bit when counting the
LSB as bit 1, or 1<<41.

The second bug is that the original bitmask value isn't right for either
description:

% ./python.exe -c "from math import log2; print(log2(0x010000000000))"
40.0

This causes the test to fail on valid MAC addresses.

Fix this by improving the comment, with references, and using the correct
bitmask.

<!-- issue-number: bpo-32107 -->
https://bugs.python.org/issue32107
<!-- /issue-number -->

warsaw committed:
bpo-32107 - Fix test_uuid

The original check was erroneous in two ways.  First, the documentation said
that "47 bit will never be set in IEEE 802 addresses obtained from network
cards".  I think this is referring to the universally vs. locally administered
MAC addresses.  Network cards from hardware manufacturers will always be
universally administered in order to guarantee global uniqueness.

But from https://en.wikipedia.org/wiki/MAC_address it says that this flag is
the "distinguished by setting the second-least-significant bit of the first
octet of the address", where a 0 means universally administered and a 1 means
it's locally administered.  This works out to the 42nd bit when counting the
LSB as bit 1, or 1<<41.

The second bug is that the original bitmask value isn't right for either
description:

% ./python.exe -c "from math import log2; print(log2(0x010000000000))"
40.0

This causes the test to fail on valid MAC addresses.

Fix this by improving the comment, with references, and using the correct
bitmask.

warsaw committed:
bpo-32107 - Fix test_uuid

The original check was erroneous in two ways.  First, the documentation said
that "47 bit will never be set in IEEE 802 addresses obtained from network
cards".  I think this is referring to the universally vs. locally administered
MAC addresses.  Network cards from hardware manufacturers will always be
universally administered in order to guarantee global uniqueness.

But from https://en.wikipedia.org/wiki/MAC_address it says that this flag is
the "distinguished by setting the second-least-significant bit of the first
octet of the address", where a 0 means universally administered and a 1 means
it's locally administered.  This works out to the 42nd bit when counting the
LSB as bit 1, or 1<<41.

The second bug is that the original bitmask value isn't right for either
description:

% ./python.exe -c "from math import log2; print(log2(0x010000000000))"
40.0

This causes the test to fail on valid MAC addresses.

Fix this by improving the comment, with references, and using the correct
bitmask.

warsaw committed:
bpo-32107 - Fix test_uuid

The original check was erroneous in two ways.  First, the documentation said
that "47 bit will never be set in IEEE 802 addresses obtained from network
cards".  I think this is referring to the universally vs. locally administered
MAC addresses.  Network cards from hardware manufacturers will always be
universally administered in order to guarantee global uniqueness.

But from https://en.wikipedia.org/wiki/MAC_address it says that this flag is
the "distinguished by setting the second-least-significant bit of the first
octet of the address", where a 0 means universally administered and a 1 means
it's locally administered.  This works out to the 42nd bit when counting the
LSB as bit 1, or 1<<41.

The second bug is that the original bitmask value isn't right for either
description:

% ./python.exe -c "from math import log2; print(log2(0x010000000000))"
40.0

This causes the test to fail on valid MAC addresses.

Fix this by improving the comment, with references, and using the correct
bitmask.

warsaw commented:
I also added a very minor simplification to the code, since it's always better to return an explicit value when there are other return paths in the code, rather than letting None be implicitly returned.

serhiy-storchaka commented on Lib/uuid.py:
```diff
@@ -404,8 +404,7 @@ def _arp_getnode():
                     lambda i: i+2)
-    if mac:
-        return mac
+    return mac
```
Should return None instead of 0.

warsaw replied:
Ah, good point.  Let me fix that.

warsaw committed:
Return None instead of 0

warsaw committed:
Return None instead of 0

warsaw committed:
Return None instead of 0

serhiy-storchaka commented on Lib/uuid.py:
```diff
@@ -404,8 +404,8 @@ def _arp_getnode():
-    if mac:
-        return mac
+    # Return None instead of 0.
+    return mac if mac else None
```
And if pep8-izy the code, I suggest to add explicit `return None` in other functions.

For uniformity I would prefer
```python
    if mac:
        return mac
    return None
```
in this function.

vstinner replied:
I like the code @serhiy-storchaka proposed.

warsaw replied:
@vstinner Yep, agreed.

vstinner committed:
bpo-31324: Optimize support._match_test() (#4421)

* Rename support._match_test() to support.match_test(): make it
  public
* Remove support.match_tests global variable. It is replaced with a
  new support.set_match_tests() function, so match_test() doesn't
  have to check each time if patterns were modified.
* Rewrite match_test(): use different code paths depending on the
  kind of patterns for best performances.

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>

warsaw committed:
Never return a locally administered MAC address

Also, clean up some return sites.

warsaw committed:
Never return a locally administered MAC address

Also, clean up some return sites.

warsaw committed:
Never return a locally administered MAC address

Also, clean up some return sites.

warsaw commented:
@serhiy-storchaka @vstinner - I think this one's got it.  The functions should never return a locally administered MAC address now.

ED: Ah darn maybe not.  Too tired.  I'll get back to this later.

vstinner commented on Lib/uuid.py:
```diff
@@ -355,7 +370,7 @@ def _find_mac(command, args, hw_identifiers, get_index):
                             word = words[get_index(i)]
                             mac = int(word.replace(b':', b''), 16)
-                            if mac:
+                            if ~(mac & (1<<41)):
```
I don't understand the usage of the ~int operator. 

What do you want to test? If a bit is unset? Do you mean "if not(mac & (1<<41)):" or "if (mac & (1<<41)) == 0:" ?

Please write a private subfunction for this test.

warsaw replied:
Yeah, that was completely botched.  Now that I've had some sleep <wink> I've fixed this with a helper function.

warsaw replied:
Also with the latest commit, we keep track of the first locally administered MAC found.  If no universally administered MAC is found, we return the first local MAC.

And the way the getters work, if there are no local or universal MACs found, then `getnode()` ultimately falls back to the RFC 4122 specified random MAC, with the multicast bit set.  So it should never be the case that `getnode()` doesn't return an RFC-defined node address.

vstinner reviewed (commented):
I understand that you want to exclude "local MAC". Ok. But what if a compute only has local MACs? uuid1() fails with a TypeError with a strange error message if getnode() returns None. At least, please modify uuid1() to raise a better error if no MAC can be found.

Or maybe fall back on a local MAC if no other address can be found?

vstinner committed:
bpo-31324: Fix test.support.set_match_tests(None) (#4505)

vstinner committed:
bpo-27535: Optimize warnings.warn() (#4508)

* Optimize warnings.filterwarnings(). Replace re.compile('') with
  None to avoid the cost of calling a regex.match() method, whereas
  it always matchs.
* Optimize get_warnings_attr(): replace PyObject_GetAttrString() with
  _PyObject_GetAttrId().

Cleanup also create_filter():

* Use _Py_IDENTIFIER() to allow to cleanup strings at Python
  finalization
* Replace Py_FatalError() with a regular exceptions

vstinner committed:
bpo-32030: Add _PyCoreConfig.module_search_path_env (#4504)

Changes:

* Py_Main() initializes _PyCoreConfig.module_search_path_env from
  the PYTHONPATH environment variable.
* PyInterpreterState_New() now initializes core_config and config
  fields
* Compute sys.path a little bit ealier in
  _Py_InitializeMainInterpreter() and new_interpreter()
* Add _Py_GetPathWithConfig() private function.

csabella committed:
bpo-32100: IDLE: Fix pathbrowser errors; improve tests. (#4484)

Patch mostly by Cheryl Sabella

Jesse-Bakker committed:
bpo-10049: Add a "no-op" (null) context manager to contextlib (GH-4464)

Adds a simpler and faster alternative to ExitStack for handling
single optional context managers without having to change the
lexical structure of your code.

vstinner committed:
bpo-32030: Move PYTHONPATH to _PyMainInterpreterConfig (#4511)

Move _PyCoreConfig.module_search_path_env to _PyMainInterpreterConfig
structure.

vstinner committed:
bpo-32030: Add _PyMainInterpreterConfig.pythonhome (#4513)

* Py_Main() now reads the PYTHONHOME environment variable
* Add _Py_GetPythonHomeWithConfig() private function
* Add _PyWarnings_InitWithConfig()
* init_filters() doesn't get the current core configuration from the
  current interpreter or Python thread anymore. Pass explicitly the
  configuration to _PyWarnings_InitWithConfig().
* _Py_InitializeCore() now fails on _PyWarnings_InitWithConfig()
  failure.
* Pass configuration as constant

xdegaye committed:
bpo-28538: Revert all the changes (now using Android Unified Headers) (GH-4479)

xdegaye committed:
bpo-29040: Support building Android with Unified Headers (GH-4492)

xdegaye committed:
bpo-28762: Revert last commit (now using Android Unified Headers) (GH-4488)

berkerpeksag committed:
bpo-1102: View.Fetch() now returns None when it's exhausted (GH-4459)

warsaw committed:
Several sleep-improved fixes

* Add a helper function to determine whether a MAC address is universally
  administered or not.
* In the various MAC getter functions, keep track of the first locally
  administered MAC, and if no universally administered MAC is found, return
  the first local MAC.
* In the ultimate fallback _random_getnode(), add a comment for the multicast
  bit being set, and use a better bitmask
* In getnode(), just fall back to _random_getnode() explicitly if no other MAC
  has been found.

warsaw committed:
Several sleep-improved fixes

* Add a helper function to determine whether a MAC address is universally
  administered or not.
* In the various MAC getter functions, keep track of the first locally
  administered MAC, and if no universally administered MAC is found, return
  the first local MAC.
* In the ultimate fallback _random_getnode(), add a comment for the multicast
  bit being set, and use a better bitmask
* In getnode(), just fall back to _random_getnode() explicitly if no other MAC
  has been found.

warsaw committed:
Several sleep-improved fixes

* Add a helper function to determine whether a MAC address is universally
  administered or not.
* In the various MAC getter functions, keep track of the first locally
  administered MAC, and if no universally administered MAC is found, return
  the first local MAC.
* In the ultimate fallback _random_getnode(), add a comment for the multicast
  bit being set, and use a better bitmask
* In getnode(), just fall back to _random_getnode() explicitly if no other MAC
  has been found.

warsaw committed:
A missed return

warsaw committed:
A missed return

warsaw committed:
A missed return

warsaw committed:
Two typos

warsaw committed:
Two typos

warsaw committed:
Two typos

vstinner committed:
bpo-32030: Rewrite calculate_path() (#4521)

* calculate_path() rewritten in Modules/getpath.c and PC/getpathp.c
* Move global variables into a new PyPathConfig structure.
* calculate_path():

  * Split the huge calculate_path() function into subfunctions.
  * Add PyCalculatePath structure to pass data between subfunctions.
  * Document PyCalculatePath fields.
  * Move cleanup code into a new calculate_free() subfunction
  * calculate_init() now handles Py_DecodeLocale() failures properly
  * calculate_path() is now atomic: only replace PyPathConfig
    (path_config) at once on success.

* _Py_GetPythonHomeWithConfig() now returns an error on failure
* Add _Py_INIT_NO_MEMORY() helper: report a memory allocation failure
* Coding style fixes (PEP 7)

vstinner committed:
bpo-27535: Cleanup create_filter() (#4516)

create_filter() now expects the action as a _Py_Identifier which
avoids string comparison, and more important, to avoid handling the
"unknown action" annoying case.

vstinner committed:
bpo-32030: Fix calculate_path() on macOS (#4526)

vstinner committed:
bpo-31979: Remove unused align_maxchar() function (#4527)

berkerpeksag committed:
bpo-19610: setup() now raises TypeError for invalid types (GH-4519)

The Distribution class now explicitly raises an
exception when 'classifiers', 'keywords' and
'platforms' fields are not specified as a list.

rhettinger committed:
bpo-32099 Add deque variant of roundrobin() recipe (#4497)



* Minor wording tweaks

berkerpeksag committed:
bpo-31325: Fix usage of namedtuple in RobotFileParser.parse() (#4529)

lelit committed:
Extending Python Doc minor updates  (GH-4518)

Move footnote markers to be closer to the related terminology:
before the end of the sentence, instead of after.

berkerpeksag committed:
bpo-12382: Make OpenDatabase() raise better exception messages (GH-4528)

Previously, 'msilib.OpenDatabase()' function raised a
cryptical exception message when it couldn't open or
create an MSI file. For example:

    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    _msi.MSIError: unknown error 6e

vstinner committed:
bpo-32096: Remove obj and mem from _PyRuntime (#4532)

bpo-32096, bpo-30860:  Partially revert the commit
2ebc5ce42a8a9e047e790aefbf9a94811569b2b6:

* Move structures back from Include/internal/mem.h to
  Objects/obmalloc.c
* Remove _PyObject_Initialize() and _PyMem_Initialize()
* Remove Include/internal/pymalloc.h
* Add test_capi.test_pre_initialization_api():
   Make sure that it's possible to call Py_DecodeLocale(), and then call
   Py_SetProgramName() with the decoded string, before Py_Initialize().

PyMem_RawMalloc() and Py_DecodeLocale() can be called again before
_PyRuntimeState_Init().

Co-Authored-By: Eric Snow <ericsnowcurrently@gmail.com>

warsaw commented:
```
>>> mac = 0x0242ac11000d
>>> mac & (1<<41)
2199023255552
```

Which means that travis is returning a locally administered MAC, which must mean it has no universally administered MACs.   This is what I was alluding to previously about environmental factors leaking into the tests.

berkerpeksag committed:
bpo-12239: Make GetProperty() return None for VT_EMPTY (GH-4539)

The previous behavior was to raise an exception

    NotImplementedError: result of type 0

when the value of the property is VT_EMPTY.

xdegaye committed:
bpo-28684: asyncio tests handle PermissionError raised on binding unix sockets (GH-4503)

The test.support.skip_unless_bind_unix_socket() decorator is used to skip
asyncio tests that fail because the platform lacks a functional bind()
function for unix domain sockets (as it is the case for non root users on the
recent Android versions that run now SELinux in enforcing mode).

Commit:
Improve the String tutorial docs (GH-4541)

The paragraph that contains example of string literal concatenation was placed
after the section about concatenation using the '+' sign.
Moved the paragraph to the appropriate section.

xdegaye committed:
bpo-28684: Remove useless import added by the previous commit (GH-4547)

vstinner committed:
bpo-32125: Remove Py_UseClassExceptionsFlag flag (#4544)

This flag was deprecated and wasn't used anymore since Python 2.0.

vstinner committed:
bpo-32124: Document C functions safe before init (#4540)

Explicitly document C functions and C variables that can be set
before Py_Initialize().

vstinner committed:
bpo-32030: Add _PyMainInterpreterConfig_ReadEnv() (#4542)

Py_GetPath() and Py_Main() now call
_PyMainInterpreterConfig_ReadEnv() to share the same code to get
environment variables.

Changes:

* Add _PyMainInterpreterConfig_ReadEnv()
* Add _PyMainInterpreterConfig_Clear()
* Add _PyMem_RawWcsdup()
* _PyMainInterpreterConfig: rename pythonhome to home
* Rename _Py_ReadMainInterpreterConfig() to
  _PyMainInterpreterConfig_Read()
* Use _Py_INIT_USER_ERR(), instead of _Py_INIT_ERR(), for decoding
  errors: the user is able to fix the issue, it's not a bug in
  Python. Same change was made in _Py_INIT_NO_MEMORY().
* Remove _Py_GetPythonHomeWithConfig()

vstinner committed:
bpo-32030: Add _PyMainInterpreterConfig.program_name (#4548)

* Py_Main() now calls Py_SetProgramName() earlier to be able to get
  the program name in _PyMainInterpreterConfig_ReadEnv().
* Rename prog to program_name
* Rename progpath to program_name

ashwoods committed:
Asyncion-Dev docs: Fix the reference to sys.excepthook (GH-4414)

vstinner committed:
bpo-32128: Skip test_nntplib.test_article_head_body() (#4552)

The NNTP server currently has troubles with SSL, whereas we don't
have the control on this server. This test blocks all CIs, so disable
it until a fix can be found.

vstinner committed:
bpo-32030: Add _PyPathConfig_Init() (#4551)

* Add _PyPathConfig_Init() and _PyPathConfig_Fini()
* Remove _Py_GetPathWithConfig()
* _PyPathConfig_Init() returns _PyInitError to allow to handle errors
  properly
* Add pathconfig_clear()
* Windows calculate_path_impl(): replace Py_FatalError() with
  _PyInitError
* Py_FinalizeEx() now calls _PyPathConfig_Fini() to release memory
* Fix _Py_InitializeMainInterpreter() regression: don't initialize
  path config if _disable_importlib is false
* PyPathConfig now uses dynamically allocated memory

mandeep committed:
bpo-30004: Fix the code example of using group in Regex Howto Docs (GH-4443)

The provided code example was supposed to find repeated words, however it returned false results.

Mariatta committed:
Improve Scheduler Objects documentation. (GH-4556)

Mention that the lower the priority number, the higher priority it represents.

berkerpeksag committed:
bpo-28334: netrc() now uses expanduser() to find .netrc file (GH-4537)

Previously, netrc.netrc() was raised an exception if $HOME is not set.

Authored-By: Dimitri Merejkowsky <dimitri.merejkowsky@tanker.io>

warsaw committed:
DEBUGGING TRAVIS - DO NOT MERGE

warsaw committed:
DEBUGGING TRAVIS - DO NOT MERGE

warsaw committed:
DEBUGGING TRAVIS - DO NOT MERGE

warsaw committed:
DEBUGGING TRAVIS - DO NOT MERGE

jonashaag committed:
bpo-32071: Add unittest -k option (#4496)

* bpo-32071: Add unittest -k option

serhiy-storchaka committed:
bpo-24641: Improved error message for JSON unserializible keys. (#4364)

Also updated an example for default() in the module docstring.
Removed quotes around type name in other error messages.

xdegaye committed:
bpo-32059: setup.py now also searches the sysroot paths (GH-4452)

detect_modules() in setup.py now also searches the sysroot paths when
cross-compiling.

xdegaye committed:
bpo-26856: Skip test_pwd on Android until issue 32033 is fixed (GH-4561)

benjaminp committed:
mark fatal_error as noreturn (#4563)

clang can't figure out that fatal_error is noreturn itself and emits warnings:

../cpython/Python/pylifecycle.c:2116:1: warning: function declared 'noreturn' should not return [-Winvalid-noreturn]
}
^
../cpython/Python/pylifecycle.c:2125:1: warning: function declared 'noreturn' should not return [-Winvalid-noreturn]
}
^

warsaw commented:
As the last commit's debugging output shows, there are only local interfaces on Travis.

ncoghlan committed:
bpo-11063: Handle uuid.h being in default include path (GH-4565)

find_file() returns an empty list if it finds the requested
header on the standard include path, so header existence
checks need to be explicitly against "is not None".

ncoghlan committed:
bpo-32096: Ensure new embedding test can find the encodings module (GH-4566)

xdegaye committed:
bpo-32126: Skip asyncio test when sem_open() is not functional (GH-4559)

warsaw committed:
Two tests cannot succeed on Travis-CI

warsaw committed:
Two tests cannot succeed on Travis-CI

warsaw committed:
Two tests cannot succeed on Travis-CI

warsaw committed:
Two tests cannot succeed on Travis-CI

warsaw commented:
I think there's nothing better that we can do except skip the two failing tests under Travis.  This last commit does that, so we'll see.

serhiy-storchaka committed:
bpo-32107: Remove the check of the multicast bit in test_uuid.

While it is required that the multicast bit (1<<40) should be set in
random generated addresses, there is no requirement that it should be
cleared in addresses obtained from network cards.

serhiy-storchaka committed:
bpo-32107: Remove the check of the multicast bit in test_uuid.

While it is required that the multicast bit (1<<40) should be set in
random generated addresses, there is no requirement that it should be
cleared in addresses obtained from network cards.

cjrh committed:
bpo-30487: automatically create a venv and install Sphinx when running make (GH-4346)

ilevkivskyi committed:
bpo-29879: Update typing documentation. (GH-4573)

- Add "version added: 3.5.2" note where it was missing.
- Remove the mention that Reversible is new in 3.5.2

daxlab committed:
bpo-23033:  Improve SSL Certificate handling (GH-937)

Wildcard is now supported in hostname when it is one and only character in
the leftmost segment.

warsaw committed:
Merge branch 'test_uuid' of github.com:warsaw/cpython into test_uuid

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

warsaw committed:
Merge branch 'test-uuid-multicast-bit' of https://github.com/serhiy-storchaka/cpython into test_uuid

This merges the relevant bits of #4572 and should be considered instead of
that PR.

warsaw commented:
Oh hells bells, the merge went badly.

warsaw committed:
Merge branch 'test_uuid' of github.com:warsaw/cpython into test_uuid

warsaw commented:
See #4576 which resolves the merge snafus.
